text/0004-api-history-schema.md

```diff
@@ -0,0 +1,249 @@
+  are necessary if the change is backported to other branches."
+  — @dsanders11 via Slack
+- The Node.js API history [includes a `removed` property](https://github.com/nodejs/node/blob/0db95d371274104a5acf09214bf8325c45bfb64a/doc/api/errors.md?plain=1#L3432-L3442),
+  however that will not be included in this schema because:
+  - When an API is removed from Electron, it is also removed from the documentation.
+    - This currently cannot be changed because [the documentation is used to generate
+    TypeScript definitions](https://github.com/electron/docs-parser/blob/d95a2a2e58d6dafce1247d4bc820ca60516fe10c/README.md?plain=1#L36-L43)
+    for the Electron API.
+- Using [Zod](https://zod.dev/) validation logic may be simpler. It's also easy
+to understand for a developer even if they've never worked with it before:
+
+  ```ts
+  const changeSchema = z.object({
+    'pr-url': z.string().url().startsWith('https://github.com/electron/electron/pull/'),
+    'breaking-changes-header': z.string().min(3).optional(),
+    description: z.string().min(3).max(72).optional(),
+  });
+
+  const historySchema = z.object({
+    added: z.array(changeSchema).optional(),
+    deprecated: z.array(changeSchema).optional(),
+    changes: z.array(changeSchema).optional(),
+  });
+  ```
+
+  - > *"Why Zod specifically?"*
+    - First-class TypeScript support and [static type inference](https://zod.dev/?id=type-inference).
+    - [Most starred](https://byby.dev/js-object-validators) JavaScript object
+    validation library.
+    - [Zero dependencies](https://www.npmjs.com/package/zod?activeTab=dependencies).
+    - Liked by developers
+    [[1]](https://www.reddit.com/r/reactjs/comments/z739b9/yup_vs_zod_what_do_you_prefer_and_why/)
+    [[2]](https://medium.com/@weidagang/zod-schema-validation-made-easy-195f86d82d44)
+    [[3]](https://rasitcolakel.medium.com/exploring-zod-a-comprehensive-guide-to-powerful-data-validation-in-javascript-typescript-2c4818b5646d).
+    - Liked in comparisons:
+    [[1]](https://zod.dev/?id=comparison)
+    [[2]](https://blog.logrocket.com/comparing-schema-validation-libraries-zod-vs-yup/#zod-vs-yup)
+    [[3]](https://www.bitovi.com/blog/comparing-schema-validation-libraries-ajv-joi-yup-and-zod).
+
+## Prior art
+
+I couldn't find an example of something exactly like this being implemented
+elsewhere.
+
+[Semantic Versioning](https://semver.org/) and
+[Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) are
+tangentially related to this concept and both *suggest* a consistent format
+for information about changes but not necessarily *enforce* it.
+
+Of course, plenty of documentation includes versioning.
+
+I think enforcing a schema on API change documentation will massively improve
+the experience of a user upgrading major versions of Electron in their project.
+The process should ideally be as effortless as possible.
+
+## Unresolved questions
```

Comment:
One interesting question is how this impacts / can be used in [`@electron/docs-parser`](https://github.com/electron/docs-parser). The module we use to to generate structured docs from markdown.

It's possible that the website we should build the parsing of this schema into that module and then have the website use `@electron/docs-parser` to do it's actual parsing. But in any case at a minimum that module will need to be updated to parse/ignore/consume the new block as we probably don't want it appearing raw in the typescript docs for instance.

Reply:
> But in any case at a minimum that module will need to be updated to parse/ignore/consume the new block as we probably don't want it appearing raw in the typescript docs for instance.

*Cough*, https://github.com/electron/docs-parser/pull/113.